src/schemes/boundary/open_boundary/mirroring.jl

```diff
@@ -0,0 +1,190 @@
+
+        for neighbor in PointNeighbors.eachneighbor(ghost_node_position,
+                                                    neighborhood_search)
+            neighbor_coords = current_coords(u_fluid, fluid_system, neighbor)
+            pos_diff = ghost_node_position - neighbor_coords
+            distance2 = dot(pos_diff, pos_diff)
+
+            if distance2 <= neighborhood_search.search_radius^2
+                distance = sqrt(distance2)
+
+                m_b = hydrodynamic_mass(fluid_system, neighbor)
+                rho_b = particle_density(v_fluid, fluid_system, neighbor)
+                pressure_b = particle_pressure(v_fluid, fluid_system, neighbor)
+                v_b = current_velocity(v_fluid, fluid_system, neighbor)
+
+                # Project `v_b` to the normal direction of the boundary zone
+                # See https://doi.org/10.1016/j.jcp.2020.110029 Section 3.3.:
+                # "Because ﬂow from the inlet interface occurs perpendicular to the boundary,
+                # only this component of interpolated velocity is kept [...]"
+                v_b = dot(v_b, boundary_zone.plane_normal) * boundary_zone.plane_normal
+
+                kernel_value = smoothing_kernel(fluid_system, distance, particle)
+                grad_kernel = smoothing_kernel_grad(fluid_system, pos_diff, distance,
+                                                    particle)
+
+                L, R = correction_arrays(kernel_value, grad_kernel, pos_diff, rho_b, m_b)
+
+                correction_matrix += L
+
+                if !prescribed_pressure
+                    interpolated_pressure += pressure_b * R
+                end
+
+                if !prescribed_velocity
+                    interpolated_velocity += v_b * R'
+                end
+            end
+        end
+
+        # See also `correction_matrix_inversion_step!` for an explanation
+        if abs(det(correction_matrix)) < 1.0f-9
+            L_inv = I
+        else
+            L_inv = inv(correction_matrix)
+        end
+
+        pos_diff = particle_coords - ghost_node_position
+
+        # In Negi et al. (2020) https://doi.org/10.1016/j.cma.2020.113119,
+        # they have modified the equation for extrapolation to
+        #
+        #           f_0 = f_k - (r_0 - r_k) ⋅ ∇f_k
+        #
+        # in order to get zero gradient at the outlet interface.
+        if prescribed_pressure
+            pressure[particle] = reference_value(reference_pressure, pressure[particle],
+                                                 particle_coords, t)
+        else
+            f_p = L_inv * interpolated_pressure
+            df_p = f_p[two_to_end]
+
+            pressure[particle] = f_p[1] + dot(pos_diff, df_p)
+        end
+
+        if prescribed_velocity
+            v_particle = current_velocity(v_open_boundary, system, particle)
+            v_ref = reference_value(reference_velocity, v_particle, particle_coords, t)
+            @inbounds for dim in 1:ndims(system)
```

Comment:
```suggestion
            @inbounds for dim in eachindex(v_ref)
```
Make this slightly safer by avoiding out-of-bounds access on `v_ref`.
Same below in the next loop.